crates/parcel_core/src/asset_graph.rs

```diff
@@ -0,0 +1,345 @@
+use std::{collections::HashSet, sync::Arc};
+
+use petgraph::{
+  graph::{DiGraph, NodeIndex},
+  visit::EdgeRef,
+  Direction,
+};
+
+use crate::types::{Asset, Dependency};
+
+#[derive(Clone, Debug)]
+pub struct AssetGraph {
+  graph: DiGraph<AssetGraphNode, AssetGraphEdge>,
+  pub assets: Vec<AssetNode>,
+  pub dependencies: Vec<DependencyNode>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct AssetNode {
+  pub asset: Asset,
+  pub requested_symbols: HashSet<String>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct DependencyNode {
+  pub dependency: Arc<Dependency>,
+  pub requested_symbols: HashSet<String>,
+  pub state: DependencyState,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum AssetGraphNode {
+  Root,
+  Entry,
+  Asset(usize),
+  Dependency(usize),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct AssetGraphEdge {}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum DependencyState {
+  New,
+  Deferred,
+  Excluded,
+  Resolved,
+}
+
+impl PartialEq for AssetGraph {
+  fn eq(&self, other: &Self) -> bool {
+    let nodes = self.graph.raw_nodes().iter().map(|n| &n.weight);
+    let other_nodes = other.graph.raw_nodes().iter().map(|n| &n.weight);
+
+    let edges = self
+      .graph
+      .raw_edges()
+      .iter()
+      .map(|e| (e.source(), e.target(), &e.weight));
+
+    let other_edges = other
+      .graph
+      .raw_edges()
+      .iter()
+      .map(|e| (e.source(), e.target(), &e.weight));
+
+    nodes.eq(other_nodes)
+      && edges.eq(other_edges)
+      && self.assets == other.assets
+      && self.dependencies == other.dependencies
+  }
+}
+
+impl AssetGraph {
+  pub fn new() -> Self {
+    let mut graph = DiGraph::new();
+
+    graph.add_node(AssetGraphNode::Root);
+
+    AssetGraph {
+      graph,
+      assets: Vec::new(),
+      dependencies: Vec::new(),
+    }
+  }
+
+  pub fn add_asset(&mut self, parent_idx: NodeIndex, asset: Asset) -> NodeIndex {
+    let idx = self.assets.len();
+
+    self.assets.push(AssetNode {
```

Comment:
Q: Why is storage split vs. just storing in the graph? 

petgraph will just have an array of nodes internally ; 

https://github.com/petgraph/petgraph/blob/master/src/graph_impl/mod.rs#L348